File: implementation/doubly_linked_list.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;

template<typename T>
class Doubly_linked_list{
private:
    struct node{
        T data;
        node * prev;
        node * next;
    };
    node * head;
    node * tail;
    // index which points to memory to last available element
    int tail_index;
public:
    Doubly_linked_list(){
        head = nullptr;
        tail = nullptr;
        tail_index = -1;
    }
    int size(){
        return tail_index+1;
    }
    bool empty(){
        return tail_index == -1;
    }
    void insert(T value,int pos){
        //if trying to insert outside the extremas of current linked list
        if(pos>tail_index+1 || pos<0){
            cout<<"Can't insert outside the bounds\n";
            raise(SIGSEGV);
        }
        else if(pos==0){
            // adding element at the head
            if(head==nullptr){
                // linked list is empty
                head=new node;
                head->data = value;
                tail_index +=1;
                return;
            }
            // linked list is not empty
            node* temp = new node;
            temp->data = value;
            temp->next = head;
            head->prev = temp;
            head = temp;
            tail_index +=1;
            return;
        }
        else if(pos==tail_index){
            // adding elements at tail
            node * temp=head;
            while(temp->next != nullptr)
                temp = temp->next;
            // create a new structure to be placed at index pos, pointer to which is 'var'
            node *var=new node;
            var->data = value;
            // add link to new data to be inserted at index 'pos' and data in linked list at index 'pos-1'
            temp->prev->next = var;
            var->prev = temp->prev;
            // add link to new data to be inserted at index 'pos' and data in current linked list at index 'pos'
            var->next = temp;
            temp->prev = var;
            
            tail_index +=1;
            return;
        }
        else if(pos==tail_index+1){
            // adding elements after tail
            node * temp = head;
            while(temp->next != nullptr)
                temp =temp->next;
            temp->data = value;
            tail_index +=1;
            return ;
        }
        else {
            // adding elements in the middle
            node * temp = head;
            int cur_index = 0;
            // temp points to structure at index pos
            while(cur_index < pos){
                cur_index += 1;
                temp = temp->next;
            }
            // create a new structure to be placed at index pos, pointer to which is 'var'
            node * var = new node;
            var->data = value;
            // add link to new data to be inserted at index 'pos' and data in linked list at index 'pos-1'
            temp->prev->next = var;
            var->prev = temp->prev;
            // add link to new data to be inserted at index 'pos' and data in current linked list at index 'pos'
            temp->next->prev = var;
            var->next = temp->next;

            tail_index +=1;
            return;
        }
    }
    T at(int pos){
        if(pos>tail_index || pos<0){
            cout<<"Can't access unavailable indexes\n";
            raise(SIGSEGV);
        }
        node *temp = head;
        int cur_index=0;
        while(cur_index!=pos){
            cur_index+=1;
            temp=temp->next;
        }
        return temp->data;
    }
};

int main(){
/*
    // Test cases :
    Doubly_linked_list<int> l;
    l.insert(23,0);
    cout<<"size of linked list is "<<l.size()<<'\n';
    l.insert(24,0);
    cout<<"size of linked list is "<<l.size()<<'\n';
    l.insert(25,1);
    cout<<"size of linked list is "<<l.size()<<'\n';

    for(int i=0;i<l.size();i++)
        cout<<"value at "<<i<<" is "<<l.at(i)<<endl;

    // l.insert(26,-1);
    // cout<<"size of linked list is "<<l.size()<<'\n';
    // l.insert(27,2);
    // cout<<"size of linked list is "<<l.size()<<'\n';
    // l.insert(28,12);
    // cout<<"size of linked list is "<<l.size()<<'\n';

*/
/*
    Causion:
    Need to handle cases where typecasting of data happens,for example,
    Doubly_linked_list<int> l;
    l.insert('a',0);
    l.insert(2.3,0);
*/
return 0;
}

File: implementation/Doubly_Linked_List/doubly_linked_list.py
# The basic class that will hold the info about another chains in the linked list
class Node:
	def __init__(self,data):
		self.prev = None
		self.next = None
		self.data = data

class Doubly_linked_list:

	def __init__(self):
		self.head = None
		self.tail_index = -1

	def size(self):
		return self.tail_index+1

	def empty(self):
		return self.tail_index is -1

	def insert(self,data,pos):
		if pos>self.tail_index+1 or pos<0:
			print("Can't insert outside the bounds")
			raise Exception("Out of bounds")
		# if inserting at head
		elif pos is 0:
			# if the list is empty
			if self.head is None:
				self.head = Node(data)
				self.tail_index += 1
				return
			# make new data point as head of list
			temp = Node(data)
			temp.next = self.head
			self.head.prev = temp
			self.head = temp
			self.tail_index += 1
			return
		# if inserting at the last position
		elif pos is self.tail_index:
			temp = self.head
			# find the last link of the list
			while temp.next is not None:
				temp = temp.next
			var = Node(data)
			temp.prev.next = var 
			var.prev = temp.prev
			var.next = temp
			temp.prev = var
			self.tail_index += 1
			return
		# if appending the data at tail
		elif pos is self.tail_index+1:
			temp = self.head
			while temp.next is not None:
				temp = temp.next
			temp.next = Node(data)
			temp.next.prev = temp
			self.tail_index += 1
			return
		# inserting somewhere in middle
		else:
			temp = self.head
			cur_index = 0
			while cur_index<pos:
				cur_index += 1
				temp = temp.next
			var = Node(data)
			temp.prev.next = var
			var.prev = temp.prev
			temp.next.prev = var
			var.next = temp.next
			self.tail_index += 1
			return

	# to erase the link at the given position
	def erase(self,pos):
		if pos > self.tail_index+1 or pos<0:
			print("can't access out of bounds")
			raise Exception("out of bounds")
		# if link at head is to be erased
		elif pos is 0:
			self.head = self.head.next
			self.tail_index = -1
			return
		# if link at tail is to be erased
		elif pos is self.tail_index:
			temp = self.head
			cur_index = 0
			while cur_index+1 < pos:
				cur_index += 1
				temp = temp.next
			temp.next = None
			self.tail_index -= 1
			return
		# if link to be erased lies somewhere in middle of list
		else:
			temp = self.head
			cur_index = 0
			while cur_index is not pos:
				cur_index += 1
				temp = temp.next
			temp.prev.next = temp.next
			temp.next.prev = temp.prev
			self.tail_index -= 1
			return

	# gives the data stored at the given position
	def at(self,pos):
		if pos > self.tail_index or pos<0:
			print("Can't access unavailable indexes")
			raise Exception("Out of bounds")
		temp = self.head
		cur_index = 0
		while cur_index is not pos:
			cur_index += 1
			temp = temp.next
		return temp.data

if __name__=="__main__":
	# print("The test cases :")
	# l = Doubly_linked_list()
	# l.insert(12,0)
	# l.insert(13,1)
	# l.insert(132,1)
	# print(l.at(1))
	# l.erase(1)